File: Inaugura.Dialogic/DialogicLine.cpp
// This is the main DLL file.

#include "stdafx.h"
#include "DialogicLegacy.h"
#include "DialogicLine.h"
#include "HangupException.h"
#include <stdio.h>

/*
 * Call Analysis termination type.
 */
#define CR_BUSY               7        /* Line busy */
#define CR_NOANS              8        /* No answer */
#define CR_NORB               9        /* No ringback */
#define CR_CNCT               10       /* Call connected */
#define CR_CEPT               11       /* Operator intercept */
#define CR_STOPD              12       /* Call analysis stopped */
#define CR_NODIALTONE         17       /* No dialtone detected */
#define CR_FAXTONE            18       /* Fax tone detected */
#define CR_ERROR              0x100    /* Call analysis error */

/*
 * Connection types ( ATDX_CONNTYPE() )
 */
#define CON_CAD               1  /* Cadence Break */
#define CON_LPC               2  /* Loop Current Drop */
#define CON_PVD               3  /* Positive Voice Detect */
#define CON_PAMD              4  /* Positive Answering Machine Detect */
#define CON_DIGITAL	

//#using <SpeechLib.dll>

//using namespace SpeechLib;

namespace Inaugura
{
namespace Dialogic
{	
	DialogicLine::DialogicLine(String __gc* channelName)
	{
		mpDialogicLine = new DialogicLegacy(channelName);		
	}

	DialogicLine::~DialogicLine(void)
	{
		delete mpDialogicLine;
	}

	// PROPERTIES 
	int DialogicLine::get_NumberOfDigitsInBuffer()
	{
		try
		{
			return mpDialogicLine->GetNumberOfDigitsInBuffer();
		}
		catch(char* str)
		{
			DialogicException __gc *ex = new DialogicException(str);
			throw ex;			
		}
	}

	String __gc* DialogicLine::get_ChannelName()
	{
		try
		{
			return (mpDialogicLine->ChannelName());
		}
		catch(char* str)
		{
			DialogicException __gc *ex = new DialogicException(str);
			throw ex;			
		}
	}

	int DialogicLine::GetHookState()
	{
		try
		{
			return mpDialogicLine->GetHookState();
		}
		catch(char* str)
		{
			DialogicException __gc *ex = new DialogicException(str);
			throw ex;			
		}
	}

	void DialogicLine::set_ChannelName(String __gc* channelName)
	{		
		try
		{
			mpDialogicLine->ChannelName(channelName);
		}
		catch(char* str)
		{
			DialogicException __gc *ex = new DialogicException(str);
			throw ex;			
		}
	}

	bool DialogicLine::get_EnableCallerId()
	{
		try
		{
			return mpDialogicLine->EnableCallerId();
		}
		catch(char* str)
		{
			DialogicException __gc *ex = new DialogicException(str);
			throw ex;			
		}
	}

	void DialogicLine::set_EnableCallerId(bool  enableCallerId)
	{	
		try
		{
			mpDialogicLine->EnableCallerId(enableCallerId);
		}
		catch(char* str)
		{
			DialogicException __gc *ex = new DialogicException(str);
			throw ex;			
		}
	}

	String __gc* DialogicLine::get_CallerId()
	{
		try
		{
			return (mpDialogicLine->GetCallerIdXml());
		}
		catch(char* str)
		{
			DialogicException __gc *ex = new DialogicException(str);
			throw ex;			
		}
	}

	// METHODS
	void DialogicLine::OpenLine()
	{
		try
		{
			mpDialogicLine->OpenLine();
		}
		catch(char* str)
		{
			DialogicException __gc *ex = new DialogicException(str);
			throw ex;			
		}
	}


	void DialogicLine::CloseLine()
	{
		try
		{
			mpDialogicLine->CloseLine();
		}
		catch(char* str)
		{
			DialogicException __gc *ex = new DialogicException(str);
			throw ex;			
		}
	}

	void DialogicLine::ClearDigitBuffer()
	{
		try
		{
			mpDialogicLine->ClearDigitBuffer();
		}
		catch(char* str)
		{
			DialogicException __gc *ex = new DialogicException(str);
			throw ex;			
		}
	}

	void DialogicLine::Dial(String __gc *phoneNumber)
	{
		try
		{
			mpDialogicLine->Dial(phoneNumber);
		}
		catch(char* str)
		{
			DialogicException __gc *ex = new DialogicException(str);
			throw ex;			
		}
	}

	CallAnalysis DialogicLine::Dial(String __gc *phoneNumber, int numberOfRings)
	{
		try
		{
			int cares = mpDialogicLine->Dial(phoneNumber,numberOfRings);
			switch(cares)
			{
				case CR_CNCT:				
					return CallAnalysis::CallConnected;
				case CR_CEPT:
					return CallAnalysis::OperatorIntercept;
				case CR_BUSY:
					return CallAnalysis::Busy;			
				case CR_FAXTONE:
					return CallAnalysis::FaxTone;			
				case CR_NODIALTONE:
					return CallAnalysis::NoDialTone;			
				case CR_STOPD:
					return CallAnalysis::CallAnalysisStopped;			
				case CR_NOANS:		
					return CallAnalysis::NoAnswer;			
				case CR_NORB:
					return CallAnalysis::NoRingBack;			
				case CR_ERROR:				
					return CallAnalysis::Error;			
			}				
		}
		catch(char* str)
		{
			DialogicException __gc *ex = new DialogicException(str);
			throw ex;			
		}
	}

	String __gc* DialogicLine::GetCallerId()
	{
		try
		{
			return mpDialogicLine->GetCallerIdXml();
		}
		catch(char* str)
		{
			DialogicException __gc *ex = new DialogicException(str);
			throw ex;			
		}
	}

	String __gc* DialogicLine::GetDigits(int maxDigits, double  timeOutSec, String __gc* termDigits, double digitTimeOutSec, double silenceTimeOutSec, String __gc* __gc* termDigit)
	{				
			int cppMaxDigits = maxDigits;
			CString cppTermDigits = termDigits;
			float cppTimeOutSec = (float)timeOutSec;
			float cppDigitTimeOutSec = (float)digitTimeOutSec;
			float cppSilenceTimeOutSec = (float)silenceTimeOutSec;
			CString cppTermDigit = (*termDigit);		
			
			String __gc* str = String::Empty;
			try
			{
				str = mpDialogicLine->GetDigits(cppMaxDigits,cppTimeOutSec,cppTermDigits,cppDigitTimeOutSec,cppSilenceTimeOutSec,cppTermDigit);
				
				(*termDigit) = cppTermDigit;
				return str;	
			}
			catch(char* exStr)
			{
				DialogicException __gc *ex = new DialogicException(exStr);
				throw ex;		
			}
			if(String::Compare(str,"HANGUP")==0)
				throw new HangupException("Hang Up Termination");		
	}

	String __gc* DialogicLine::GetDigits(int maxDigits, double  timeOutSec, String __gc* termDigits, double digitTimeOutSec)
	{		
		String __gc* returnString = new String("");
		String __gc* str = GetDigits(maxDigits,timeOutSec,termDigits,digitTimeOutSec,-1,&returnString);			
		return str;	
	}

	String __gc* DialogicLine::GetDigits(int maxDigits, double  timeOutSec, double digitTimeOutSec)
	{		
		String __gc* returnString = new String("");
		String __gc* str = GetDigits(maxDigits,timeOutSec,"",digitTimeOutSec,-1,&returnString);			
		return str;	
	}

	int DialogicLine::DetectRingback(double timeOutSec)
	{				
			float cppTimeOutSec = (float)timeOutSec;			
			int result = 0;
			try
			{
				result = mpDialogicLine->DetectRingback(cppTimeOutSec);
				return result;
			}
			catch(char* exStr)
			{
				DialogicException __gc *ex = new DialogicException(exStr);
				throw ex;		
			}
	}


	void DialogicLine::GiveDialTone()
	{
		try
		{
			mpDialogicLine->GiveDialTone();			
		}
		catch(char* str)
		{
			DialogicException __gc *ex = new DialogicException(str);
			throw ex;			
		}
	}

	void DialogicLine::OnHook()
	{
		try
		{
			mpDialogicLine->OnHook();			
		}
		catch(char* str)
		{
			DialogicException __gc *ex = new DialogicException(str);
			throw ex;			
		}
	}

	void DialogicLine::OffHook()
	{
		try
		{
			mpDialogicLine->OffHook();			
		}
		catch(char* str)
		{
			DialogicException __gc *ex = new DialogicException(str);
			throw ex;			
		}
	}

	int DialogicLine::PlaySpecial(String __gc* fileName)
	{
		CString str2 = fileName;
		int result = 0;
		try
		{
			result = mpDialogicLine->PlaySpecial(str2);			
		}
		catch(char* str)
		{
			DialogicException __gc *ex = new DialogicException(str);
			throw ex;			
		}			
		catch(Exception __gc *ex)
		{
			DialogicException __gc *dialogicEx = new DialogicException("Exception thrown by internal component",ex);
			throw dialogicEx;			
		}
		
		if(result == HANGUP)
			throw new HangupException("Hang Up Termination");
		else if(result == TONETERM)
			throw new HangupException("Hang Up via tone detection");
		
		return result;
	}

	int DialogicLine::PlayFile(String __gc* fileName, String __gc* termDigits, bool terminate)
	{
		CString str2 = fileName;
		int result = 0;
		try
		{
			result = mpDialogicLine->PlayFile(str2,termDigits,terminate?1:0);			
		}
		catch(char* str)
		{
			DialogicException __gc *ex = new DialogicException(str);
			throw ex;			
		}			
		catch(Exception __gc *ex)
		{
			DialogicException __gc *dialogicEx = new DialogicException("Exception thrown by internal component",ex);
			throw dialogicEx;			
		}
		
		if(result == HANGUP)
			throw new HangupException("Hang Up Termination");			
		else
		{
			if(result == DIGITTERM)
				return 1;
			else 
				return 0;
		}				
	}

	void DialogicLine::PlayFile(String __gc* fileName)
	{
		this->PlayFile(fileName,"",false);		
	}

	void DialogicLine::RecordWav(String __gc* fileName, double  timeOutSec, double  silenceTimeOutSec, bool terminate )
	{
		int result = 0;
		try
		{
			result = mpDialogicLine->RecordFile(fileName,timeOutSec,silenceTimeOutSec,terminate?1:0);
		}
		catch(char* str)
		{			
			throw new DialogicException(str);				
		}
		catch(Exception __gc *ex)
		{
			DialogicException __gc *dialogicEx = new DialogicException("Exception thrown by internal component",ex);
			throw dialogicEx;			
		}		
		if(result == HANGUP)
			throw new HangupException("Hang Up Termination");
	}

	/*
	int DialogicLine::PlayText(String __gc* text, String __gc* termDigits, bool terminate)
	{
		SpVoiceClass *sp = NULL;
		SpWaveFormatEx *wf = NULL;
		SpAudioFormatClass *sf = NULL;
		SpMemoryStreamClass *buffer = NULL;

		try
		{
			SpVoiceClass *sp = new SpVoiceClass();			
			SpWaveFormatExClass *wf = new SpWaveFormatExClass();			
			wf->set_BitsPerSample(8);
			wf->set_Channels(1);
			wf->set_SamplesPerSec(11000);
			wf->set_BlockAlign(1);
			wf->set_AvgBytesPerSec(11000);
			wf->set_FormatTag(0x0007);

			SpAudioFormatClass *sf = new SpAudioFormatClass();
			sf->SetWaveFormatEx(wf);

			SpeechVoiceSpeakFlags SpFlags = SpeechVoiceSpeakFlags::SVSFlagsAsync;
			SpeechStreamFileMode SpFileMode = SpeechStreamFileMode::SSFMCreateForWrite;
			SpMemoryStreamClass *buffer = new SpMemoryStreamClass();
			buffer->set_Format(sf);
			sp->set_AudioOutputStream((ISpeechBaseStream __gc *)buffer);
			sp->Speak(text, SpFlags);
			sp->WaitUntilDone(-1);
			System::Object __gc* data = buffer->GetData();
			System::Array * array = (System::Array *)data;			
   
			unsigned char *buf = (unsigned char *)malloc(array->get_Length() * sizeof(unsigned char));
			FILE *fp = fopen("test.wav","wb");			
			for(int i = 0; i < array->get_Length(); i++)
			{   				
				buf[i] = System::Byte::Parse(array->get_Item(i)->ToString());							
				fwrite(&buf[i],sizeof(unsigned char),1,fp);
			}
			fclose(fp);

			SpFileStreamClass __gc* SpFileStream = new SpFileStreamClass();
			SpFileStream->set_Format(sf);
			SpFileStream->Open("working.wav", SpFileMode, false);
			sp->set_AudioOutputStream((ISpeechBaseStream __gc *)SpFileStream);
			sp->Speak(text, SpFlags);
			sp->WaitUntilDone(-1);
			SpFileStream->Close();
            
			int result = mpDialogicLine->PlayFile("working.wav",termDigits,terminate?1:0);

			if(result == HANGUP)
				throw new DialogicHangupException("Hang Up Termination");			
			else
			{
				if(result != DIGITTERM)
					return 1;
				else
					return 0;
			}				
		}
		catch(char* str)
		{
			ApplicationException __gc *ex = new ApplicationException(str);
			throw ex;			
		}		
	}

	void DialogicLine::PlayText(String __gc* text)
	{
		this->PlayText(text,"",false);		
	}
	*/
	
	bool DialogicLine::WaitRing(int numberOfRings, int waitTimer)
	{
		bool result = false;
		try
		{
			result = mpDialogicLine->WaitRing(numberOfRings,waitTimer);
		}
		catch(char* str)
		{
			DialogicException __gc *ex = new DialogicException(str);
			throw ex;			
		}
			
		if(result && this->get_EnableCallerId())
		{
			try
			{
				mpDialogicLine->GetCallerId();			
			}
			catch(char* str)
			{
				DialogicException __gc *ex = new DialogicException(str);
				throw ex;			
			}							
			return result;					
		}

		return false;
	}

	float DialogicLine::TimeOut(int numberOfDigits)
	{
		return numberOfDigits * DialogicLine::TimeOutPerDigit;
	}	

	String __gc* DialogicLine::ToString()
	{
		try
		{
			return mpDialogicLine->ChannelName();
		}
		catch(char* str)
		{
			DialogicException __gc *ex = new DialogicException(str);
			throw ex;			
		}
	}
}
}

File: Inaugura.Telephony.Dialogic/DialogicLine.cs
using System;
using System.Xml;

using Inaugura.Dialogic;
using Inaugura.Telephony;


using System.Diagnostics;

namespace Inaugura.Telephony.Dialogic
{
	/// <summary>
	/// 
	/// </summary>
	public class DialogicLine : TelephonyLine
	{
		#region Variables
		protected Inaugura.Dialogic.DialogicLine mLine;
		#endregion

		#region Properties
		public override int NumberOfDigitsInBuffer
		{
			get
			{
				return this.mLine.NumberOfDigitsInBuffer;
			}
		}

		public override CallerID CallerID
		{
			get
			{
                try
                {
                    string callerIdXml = this.mLine.CallerId;

                    if (callerIdXml == null || callerIdXml == string.Empty)
                        return new CallerID("NOTSET", "", "", "", DateTime.Now, "");

                    //Log.AddLog(callerIdXml);

					XmlDocument xmlDoc = new XmlDocument();
					xmlDoc.LoadXml(mLine.CallerId);

					XmlElement node = (XmlElement)xmlDoc.FirstChild;

                    string status = "";
                    string frameType = "";
                    string name = "";
                    string number = "";
                    string dialedNumber = "";
                    string strDateTime = "";

					if (node != null)
					{
						DateTime dateTime = DateTime.Now;

						status = node.GetAttribute("status");
						frameType = node.GetAttribute("frameType");
						name = node.GetAttribute("sourceName");
						number = node.GetAttribute("sourceNumber");
						dialedNumber = node.GetAttribute("dialedNumber");
						strDateTime = node.GetAttribute("dateTime");

						if (strDateTime != string.Empty)
						{
							int month = int.Parse(strDateTime.Substring(0, 2));
							int day = int.Parse(strDateTime.Substring(2, 2));
							int hour = int.Parse(strDateTime.Substring(4, 2));
							int minute = int.Parse(strDateTime.Substring(6, 2));
							dateTime = new DateTime(dateTime.Year, month, day, hour, minute, 0, 0);
						}
						return new CallerID(status, name, number, dialedNumber, dateTime, frameType);
					}

					return null;

				}
                catch (DialogicException ex)
                {
					throw new TelephonyException(ex.Message,ex);
                }                
            }
		}	

		public override HookState HookState
		{
			get
			{
				try
				{
					if (this.mLine.GetHookState() == 1)
						return HookState.OnHook;
					else
						return HookState.OffHook;
				}
				catch (DialogicException ex)
				{
					throw new TelephonyException(ex.Message, ex);
				}
			}
		}
		#endregion

		#region Methods
        /// <summary>
        /// Clears all digits from the digit buffer
        /// </summary>
		public override void ClearDigitBuffer()
		{
			try
			{
                
                mLine.ClearDigitBuffer();
			}
			catch (DialogicException ex)
			{
				throw new TelephonyException(ex.Message, ex);
			}
		}
			
		public override void Dial(string  phoneNumber)
		{
			if(phoneNumber == null)
				throw new ArgumentNullException("phoneNumber","The phone number can not be null");

			try
			{
				mLine.Dial(phoneNumber);
			}
			catch(Inaugura.Dialogic.HangupException ex)
			{
				throw new HangupException(ex.Message,ex);
			}
			catch (DialogicException ex)
			{
				throw new TelephonyException(ex.Message, ex);
			}
		}

		public override DialAnalysis Dial(string  phoneNumber, int numberOfRings)
		{
			if (phoneNumber == null)
				throw new ArgumentNullException("phoneNumber", "The phone number can not be null");
			else if (numberOfRings <= 0)
				throw new ArgumentOutOfRangeException("numberOfRings", "numberOfRings must be greater then zero");

			CallAnalysis c = CallAnalysis.Error;

			try
			{
				c = mLine.Dial(phoneNumber,numberOfRings);
			}
			catch(Inaugura.Dialogic.HangupException ex)
			{
				throw new HangupException(ex.Message,ex);
			}
			catch (DialogicException ex)
			{
				throw new TelephonyException(ex.Message, ex);
			}

			switch(c)
			{
				case CallAnalysis.Busy:
					return DialAnalysis.Busy;
				case CallAnalysis.OperatorIntercept:
					return DialAnalysis.OperatorIntercept;				
				case CallAnalysis.CallConnected:
					return DialAnalysis.CallConnected;
				case CallAnalysis.Error:
					return DialAnalysis.Error;
				case CallAnalysis.FaxTone:
					return DialAnalysis.FaxTone;
				case CallAnalysis.NoAnswer:
					return DialAnalysis.NoAnswer;
				case CallAnalysis.NoDialTone:
					return DialAnalysis.NoDialTone;
				case CallAnalysis.NoRingBack:
					return DialAnalysis.NoRingBack;
				default:
					return DialAnalysis.Error;					
			}
		}

		#region GetDigits
		public override string GetDigits(int maxDigits, TimeSpan timeOut, TimeSpan digitTimeOut, string termDigits, out string termDigit)
		{			
			if (termDigits == null)
				termDigits = String.Empty;
			try
			{
				string termDigitTemp = string.Empty;
				string result = mLine.GetDigits(maxDigits, timeOut.TotalSeconds, termDigits, digitTimeOut.TotalSeconds, 0, ref termDigitTemp);
				termDigit = termDigitTemp;
				return result;
			}
			catch (Inaugura.Dialogic.HangupException ex)
			{
				throw new HangupException(ex.Message, ex);
			}
			catch (DialogicException ex)
			{
				throw new TelephonyException(ex.Message, ex);
			}
		}

		#endregion

		#region PlayFile
		public override void PlaySpecial(string fileName)
		{
			try
			{
				//Log.AddLog(string.Format("Playing(0): {0}", fileName));
				int result = mLine.PlaySpecial(fileName);
				Log.AddLog(string.Format("Result of playing file (with tone detection) = {0}", result));
			}
			catch (Inaugura.Dialogic.HangupException ex)
			{
				throw new HangupException(ex.Message, ex);
			}
			catch (DialogicException ex)
			{
				//Log.AddError(ex.ToString());
				throw new TelephonyException(string.Format("Error playing '{0}'", fileName), ex);
			}
			catch
			{
				//Log.AddError("Some other exception");
				throw new Exception("Some other exception");
			}
		}

		public override void PlayFile(string fileName)
		{
			try
			{
				//Log.AddLog(string.Format("Playing(0): {0}", fileName));
				mLine.PlayFile(fileName);
			}
			catch (Inaugura.Dialogic.HangupException ex)
			{
				//Log.AddError(ex.ToString());
				throw new HangupException(ex.Message, ex);
			}
			catch (DialogicException ex)
			{
				//Log.AddError(ex.ToString());
				throw new TelephonyException(string.Format("Error playing '{0}'", fileName), ex);
			}
			catch
			{
				//Log.AddError("Some other exception");
				throw new Exception("Some other exception");
			}
		}
		
		public override TelephonyResult PlayFile(string fileName, bool terminate)
		{
			try
			{
				//Log.AddLog(string.Format("Playing(1): {0}", fileName));
				if (terminate)
				{
					//Log.AddLog("Here");
					int result = mLine.PlayFile(fileName, "1234567890*#", true);
					//Log.AddLog(string.Format("Result: {0}",result));
					if (result == 1)
					{
						//Log.AddLog("Here 4");
						return TelephonyResult.Terminated;
					}
					else
					{
						//Log.AddLog("Here 5");
						return TelephonyResult.Completed;
					}
				}
				else
				{
					mLine.PlayFile(fileName);
					return TelephonyResult.Completed;
				}
			}
			catch (Inaugura.Dialogic.HangupException ex)
			{
				//Log.AddLog("Here 1");
				//Log.AddError(ex.ToString());
				throw new HangupException(ex.Message, ex);
			}
			catch (DialogicException ex)
			{
				//Log.AddLog("Here 2");
				//Log.AddError(ex.ToString());
				throw new TelephonyException(string.Format("Error playing '{0}'",fileName), ex);
			}
			catch
			{
				//Log.AddLog("Here 3");
				//Log.AddError("Some other exception");
				throw new Exception("Some other exception");
			}
		}

		public override TelephonyResult PlayFile(string fileName, string termDigits, out string termDigit)
		{
			try
			{
				termDigit = string.Empty;
				//Log.AddLog(string.Format("Playing(2): {0}", fileName));
				int result = mLine.PlayFile(fileName, termDigits, false);
				//Log.AddLog(string.Format("Result: {0}", result));
				if (result == 0) // no term digit
					return TelephonyResult.Completed;
				else // terminated because of a digit
				{
					string refTermDigit = string.Empty;
					mLine.GetDigits(mLine.NumberOfDigitsInBuffer, 0, termDigits, 0, 0, ref refTermDigit);
					termDigit = refTermDigit;
					return TelephonyResult.Terminated;
				}
			}
			catch (Inaugura.Dialogic.HangupException ex)
			{
				//Log.AddError(ex.ToString());
				throw new HangupException(ex.Message, ex);
			}
			catch (DialogicException ex)
			{
				//Log.AddError(ex.ToString());
				throw new TelephonyException(string.Format("Error playing '{0}'", fileName), ex);
			}
			catch
			{
				//Log.AddError("Some other exception");
				throw new Exception("Some other exception");
			}
		}	
		#endregion

		#region PlayText
		/*
		public override void PlayText(string voice, string text)
		{
			string fileName = this.Name;
			try
			{				
				Inaugura.Speech.Engine.SaveWav(voice,text,fileName+".wav");								
				mLine.PlayFile(fileName+".wav");				
			}
			catch(DialogicHangupException ex)
			{
				throw new HangupException(ex.Message);
			}			
		}

		public override TelephonyResult PlayText(string voice, string text, bool terminate)
		{
			string fileName = this.Name;
			try
			{				
				Inaugura.Speech.Engine.SaveWav(voice,text,fileName+".wav");				
				if(mLine.PlayFile(fileName+".wav","",terminate) == 1)
					return TelephonyResult.Completed;
				else
					return TelephonyResult.Terminated;
			}
			catch(DialogicHangupException ex)
			{
				throw new HangupException(ex.Message);
			}			
		}

		public override TelephonyResult PlayText(string voice, string text, string termDigits, ref string termDigit)
		{
			string fileName = this.Name;
			try
			{				
				Inaugura.Speech.Engine.SaveWav(voice,text,fileName+".wav");				
				if(mLine.PlayFile(fileName+".wav",termDigits,true) == 1)
					return TelephonyResult.Completed;
				else
				{
					termDigit = mLine.GetDigits(1,0,0);
					return TelephonyResult.Terminated;
				}
			}
			catch(DialogicHangupException ex)
			{
				throw new HangupException(ex.Message);
			}			
		}
		*/
		#endregion		

		public override void GiveDialTone()
		{
			mLine.GiveDialTone();
		}

		public override void OffHook()
		{
			if(this.HookState == HookState.OnHook)
				mLine.OffHook();
		}

		public override void OnHook()
		{
			if(this.HookState == HookState.OffHook)			
				mLine.OnHook();
		}

		public override void OpenLine()
		{	
			mLine.OpenLine();							
		}

		public override void CloseLine()
		{
			mLine.CloseLine();						
		}

		public override void Record(string fileName, TimeSpan timeOut, TimeSpan silenceTimeOut, bool terminate)
		{
			try
			{
				mLine.RecordWav(fileName,timeOut.TotalSeconds,silenceTimeOut.TotalSeconds,terminate);		
			}
			catch (Inaugura.Dialogic.HangupException ex)
			{
				throw new HangupException(ex.Message, ex);
			}
			catch (DialogicException ex)
			{
				throw new TelephonyException(ex.Message, ex);
			}
		}

		/*
		public override string Reco()
		{
			string str = "";
			try
			{
				this.Record(this.Name+".wav");
				str = Inaugura.Speech.Engine.Reco(this.Name+".wav");
				
			}
			catch (Inaugura.Dialogic.HangupException ex)
			{
				throw new HangupException(ex.Message, ex);
			}
			catch (DialogicException ex)
			{
				throw new TelephonyException(ex.Message, ex);
			}

			return str;			
		}
				
		public override string Reco(string grammerXml)
		{
			string str = "";
			try
			{
				this.Record(this.Name+".wav");
				str = Inaugura.Speech.Engine.Reco(this.Name+".wav",grammerXml);
				
			}
			catch (Inaugura.Dialogic.HangupException ex)
			{
				throw new HangupException(ex.Message, ex);
			}
			catch (DialogicException ex)
			{
				throw new TelephonyException(ex.Message, ex);
			}

			return str;			
		}
		*/



		public override bool WaitRing(int numberOfRings, int waitTimer)
		{
			return mLine.WaitRing(numberOfRings,waitTimer);
		}

		public DialogicLine(TelephonyHardware hardware, Inaugura.Dialogic.DialogicLine line) : base(hardware, line.ChannelName)
		{
			mLine = line;
		}

		public override bool DetectRingback(TimeSpan timeOut)
		{
			Log.AddLog(string.Format("Attempting to detect ringback for the next {0} seconds",timeOut.TotalSeconds));
			int result = mLine.DetectRingback(timeOut.TotalSeconds);
			if (result == (int)Inaugura.Dialogic.Termination.Tone)
			{
				Log.AddLog(string.Format("Result of ringback detection = Tone Detected", result));
				return true;
			}
			else
			{
				Log.AddLog(string.Format("Result of ringback detection = {0}", result));
				return false;
			}
		}
		#endregion				
	}
}
